auction-server/src/auction/service/simulator.rs

```diff
@@ -0,0 +1,336 @@
+                    if let Ok(UpgradeableLoaderState::Program {
+                        programdata_address,
+                    }) = account.state()
+                    {
+                        programs_to_fetch.push((*account_key, programdata_address));
+                    }
+                } else if account.executable {
+                    if account.owner == solana_sdk::bpf_loader::id() {
+                        accounts_config
+                            .programs
+                            .insert(*account_key, account.clone());
+                    }
+                } else {
+                    accounts_config
+                        .accounts
+                        .insert(*account_key, account.clone());
+                }
+            }
+        }
+
+        let indirect_keys = programs_to_fetch
+            .iter()
+            .map(|(_, programdata_address)| *programdata_address)
+            .collect::<Vec<_>>();
+        let indirect_accounts = self
+            .get_multiple_accounts_with_cache(&indirect_keys)
+            .await?;
+        for ((program_key, _), program_account) in
+            programs_to_fetch.iter().zip(indirect_accounts.iter())
+        {
+            if let Some(program_account) = program_account {
+                accounts_config
+                    .upgradable_programs
+                    .insert(*program_key, program_account.clone());
+            }
+        }
+
+        Ok(Response {
+            value:   accounts_config,
+            context: accounts_with_context.context,
+        })
+    }
+
+    pub async fn simulate_transaction(
+        &self,
+        transaction: &VersionedTransaction,
+    ) -> RpcResult<Result<SimulatedTransactionInfo, FailedTransactionMetadata>> {
+        let pending_txs = self.fetch_pending_and_remove_old_txs().await;
+        let txs_to_fetch = pending_txs
+            .iter()
+            .chain(std::iter::once(transaction))
+            .cloned()
+            .collect::<Vec<_>>();
+        let accounts_config_with_context = self.fetch_tx_accounts_via_rpc(&txs_to_fetch).await?;
+
+
+        let accounts_config = accounts_config_with_context.value;
+        let mut svm = LiteSVM::new()
+            .with_sigverify(false)
+            .with_blockhash_check(false)
+            .with_transaction_history(0);
+        // this is necessary for correct lookup table access
+        // otherwise 0 = slot < table.last_extended_slot
+        svm.warp_to_slot(accounts_config_with_context.context.slot);
+        accounts_config.apply(&mut svm);
+
+        pending_txs.into_iter().for_each(|tx| {
+            let _ = svm.send_transaction(tx);
+        });
+        let res = svm.simulate_transaction(transaction.clone());
+        Ok(Response {
+            value:   res,
+            context: accounts_config_with_context.context,
+        })
+    }
+
+    pub async fn optimize_bids(&self, bids: &[Bid<Svm>]) -> RpcResult<Vec<Bid<Svm>>> {
```

Comment:
We should not call this concurrently for the same auction conclusion. I think rn we have a lock for each auction. Maybe leave a comment here :-?

Reply:
Do you mean we shouldn't run two auctions for the same permission key at the same time?

Reply:
Yes. We have a lock for handle_auction. so it's fine